### ISSUE TITLE
Race condition in Azure Service Bus queue

### ISSUE THREAD

**Issue by jamie94bc** [exceptionless/Foundatio#43]
As mentioned by @moswald in #38 ...

1. In parallel, Thread A calls `EnqueueAsync()` and Thread B calls `DeleteQueueAsync()`
2. Thread A passes through `EnsureQueueCreatedAsync()`, ensuring `_queueClient` is not null
3. Thread B deletes the queue, setting `_queueClient` to null
4. Thread A tries to use `_queueCleint` which is now null

There's not really a nice way to handle this without locking. Couple of solutions:

#### Option 1 #####
Wrap `_queueClient` in a property which throws a slightly different exception, leaving the user to retry.

#### Option 2 ####
Implement retry logic (preferably in QueueBase).

**Comment by niemyjski** [comment]
Could use the new `?` null check keyword which will help prevent against this too. I'm trying to think of a scenario when you would ever want to delete a queue..

**Comment by jamie94bc** [comment]
@niemyjski yeah I'm with you on that. I'd rather replace `DeleteQueueAsync()`

**Comment by jamie94bc** [comment]
`?` would silently fail, which wouldn't be fantastic behavior!

**Comment by niemyjski** [comment]
Is this still an issue?

**Comment by jamie94bc** [comment]
Yeah - it's different from #45 (another race condition in `AzureServiceBusQueue`)

**Comment by moswald** [comment]
It is an issue, but I don't know if it can be solved without using locks everywhere. I think the "correct" solution is to document that it's a dangerous method and shouldn't be used in multi-threaded environments. Once one part of your application determines the queue should be deleted, the rest of the application should be halted.

**Comment by niemyjski** [comment]
@moswald would you mind updating the docs for this?

**Comment by niemyjski** [comment]
I'm going to close this as we've mitigated the main race conditions in the latest client. However, there are still some limited gaps around deletion but considering that is a lifecycle operation, it feels like we shouldn't carry all of the overhead of locking for an action that doesn't happen very often. Please let us know if if this is ok, or if you have any other feedback.
